### ISSUE TITLE
ContourSeries transpose is not working

### ISSUE THREAD

**Issue by objorke** [oxyplot/oxyplot#807]
similar to #535

**Comment by VisualMelon** [comment]
This is also true of Area, Line, and many others. I would propose that we move the Transform and Inverse transform shadow methods (as seen in the HeatMapSeries) into the XYSeries class. I have tested this and it appears to work for the Contour Series, though there are issues with other series (doesn't make a lot of sense for Column or Bar), but this shouldn't break anything, and issues with individual series can be addressed as they are discovered.

I'll put together a pull request if this approach seems reasonable, moving the Transform, InverseTransform, and ClippingPlane methods from the HeatMapSeries to the XYSeries.

**Comment by objorke** [comment]
How will this affect performance? The transform method could be a hot spot... I am concerned about adding this code for a special case that will affect all users of the library...

The alternative could be to require the x-axis to be horizontal! Then the application would be responsible for transforming (orientate) its data points or data arrays...

**Comment by VisualMelon** [comment]
If performance is a concern, perhaps support for carrying out transpositions at a higher level would be suitable, having this check/computation occur only once, rather than during every rendering. I would expect the cost associated with this to be lost in amid the rest of the transformation stuff, and to be insignificant compared to the rendering.

It would seem odd to disallow vertical x-axes given the existing support, but since none of the transpositions really work, I can't image it's in demand much.

Another idea that has just occurred to me and doesn't look infeasible, would be to have a dedicated vertical x-axis/horizontal y-axis class (TransposedLinear/LogAxis), which sort out the swap internally (since all the Transform/ClippingRect methods end up consulting the X Axis anyhow). This might mean that you'd need one of these for each type of axis, but it would remove the performance concern (for non-transposed plots, at least). A no doubt less performance approach might be to have a generic wrapper class for Axis which sits on-top, and fiddles with the appropriate methods to facilitate the transposition.

(I'm certainly interested in working on a solution should one be deemed appropriate, and will duly tidy up the HeatMapSeries if applicable.)

**Comment by objorke** [comment]
I think it is ok to perform the check at every rendering, but not at every transform.
Could we make a method in XYSeries that transforms a sequence of points? 

It would be great if we could find a way to support vertical x-axes, even if it is not in frequent use.

I don't like the solution with TransposedLinearAxis etc. This increases complexity, and will be difficult to explain, I think.

**Comment by VisualMelon** [comment]
I'd agree, more axes isn't going to be nice to deal with.

Provisioning the transformation of a sequence of points would require substantial changes to some of the series classes to take effect, though I may not have understood your intent here...

I think the tidiest solution would be to perform the check in the Update call of some variety (in the XYSeries, perhaps), and set a delegate up to carry out the transformation. A concern with this, however, is that it might prevent inlining and result in extra method calls, but since the Axis.Transform methods are all virtual, I doubt it. I'll investigate a solution that involves the Axis class exposing Transform & InverseTransform properties (delegates), which are resolved ahead of time, or perhaps refitting the Transform of the XYSeries class.

**Comment by objorke** [comment]
Yes, please investigate use of delegate properties for Transform & InverseTransform! It sounds like a good idea! 

Maybe the https://github.com/oxyplot/oxyplot/tree/develop/Source/Examples/PerformanceTest example can be extended for such profiling. I used it for profiling model update and rendering in the past.
